streaming/base/util.py

```diff
@@ -209,6 +215,179 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
+
+            The pattern of index_file_urls and corresponding reaction is one of:
+            1. All urls are str (local). All urls are accessible locally -> no download
+            2. All urls are tuple (local, remote). All urls are accessible locally -> no download
+            3. All urls are tuple (local, remote). At least one url is not accessible locally -> download all
+            4. All urls are str (remote) -> download all
+
+        out (Union[str, Tuple[str, str]]): path to put the merged index file
+        keep_local (bool): Keep local copy of the merged index file. Defaults to ``True``
+        download_timeout (int): The allowed time for downloading each json file. Defaults to 60s.
+    """
+    from streaming.base.storage.download import download_file
+    from streaming.base.storage.upload import CloudUploader
+
+    if not index_file_urls or not out:
+        logger.warning('Need to specify both index_file_urls and out. No index merged')
+        return
+
+    # This is the index json file name, e.g., it is index.json as of 0.6.0
+    index_basename = get_index_basename()
+
+    cu = CloudUploader.get(out, keep_local=True, exist_ok=True)
+
+    # Remove duplicates, and strip '/' from right if any
+    index_file_urls = list(set(index_file_urls))
+    urls = []
+    for url in index_file_urls:
+        if isinstance(url, str):
+            urls.append(url.rstrip('/').strip())
+        else:
+            urls.append((url[0].rstrip('/').strip(), url[1].rstrip('/').strip()))
+
+    # Prepare a temp folder to download index.json from remote if necessary. Removed in the end.
+    with tempfile.TemporaryDirectory() as temp_root:
+        logging.warning(f'Create a temporary folder {temp_root} to store index files')
+
+        # Copy files to a temporary directory. Download if necessary
+        partitions = []
+        for url in urls:
+            if isinstance(url, tuple):
+                src = url[0] if os.path.exists(url[0]) else url[1]
+            else:
+                src = url
+
+            obj = urllib.parse.urlparse(src)
+            scheme, bucket, path = obj.scheme, obj.netloc, obj.path
+            if scheme == '' and bucket == '' and path == '':
+                raise FileNotFoundError(
+                    f'Check data availability! local index {url[0]} is not accessible.' +
+                    'remote index {url[1]} does not have a valid url format')
+            dest = os.path.join(temp_root, path.lstrip('/'))
+
+            try:
+                download_file(src, dest, download_timeout)
+            except Exception as ex:
+                raise RuntimeError(f'Failed to download index.json: {src} to {dest}') from ex
+
+            if not os.path.exists(dest):
+                raise FileNotFoundError(f'Index file {dest} does not exist or not accessible.')
+
+            partitions.append(dest)
+
+        # merge index files into shards
```

Comment:
```suggestion
        # merge shards from all index files
```

Reply:
iiuc this is what this section is doing right? just wanted it to be clear for someone else going thru the function

Reply:
Yes, good catch! Changed.